File: ChanSentry.Cli/Services/ThreadFetchService.cs
using ChanSentry.CLI.Services.Interfaces;
using ChanSentry.Common.Helpers;
using ChanSentry.Common.Models;
using Spectre.Console;

namespace ChanSentry.CLI.Services;

public class ThreadFetchService : IThreadFetchService
{
    private const string UserAgent = "ChanSentry/1.0";

    public async Task<ThreadFetchResult> FetchThreadAsync(WatchedThread thread)
    {
        using var httpClient = CreateHttpClient(thread.LastChecked);
        
        var url = string.Format(Common.Constants.Urls.ThreadUrlTemplate, thread.Board, thread.ThreadId);
        var response = await httpClient.GetAsync(url);

        if (response.IsSuccessStatusCode)
        {
            var content = await response.Content.ReadAsStringAsync();
            var threadData = JsonHelper.Deserialize<Common.Models.Thread>(content);
            return ThreadFetchResult.Success(threadData);
        }

        if (response.StatusCode == System.Net.HttpStatusCode.NotModified)
        {
            return ThreadFetchResult.NotModified();
        }

        return ThreadFetchResult.Failed(response.StatusCode);
    }

    private static HttpClient CreateHttpClient(DateTime lastChecked)
    {
        var httpClient = new HttpClient();
        httpClient.DefaultRequestHeaders.Add("User-Agent", UserAgent);
        httpClient.DefaultRequestHeaders.Add("If-Modified-Since", lastChecked.ToString("R"));
        return httpClient;
    }
}

public class ThreadFetchResult
{
    public bool IsSuccess { get; init; }
    public bool IsNotModified { get; init; }
    public Common.Models.Thread? ThreadData { get; init; }
    public System.Net.HttpStatusCode? StatusCode { get; init; }

    public static ThreadFetchResult Success(Common.Models.Thread threadData)
        => new() { IsSuccess = true, ThreadData = threadData };

    public static ThreadFetchResult NotModified()
        => new() { IsNotModified = true };

    public static ThreadFetchResult Failed(System.Net.HttpStatusCode statusCode)
        => new() { StatusCode = statusCode };
}

File: src/thread-fetch-service.js
import { URLS, USER_AGENT } from './constants.js';
import { createCatalogThreads, createThread } from './models.js';

export class ThreadFetchResult {
  constructor({ isSuccess = false, isNotModified = false, threadData = null, statusCode = null } = {}) {
    this.isSuccess = isSuccess;
    this.isNotModified = isNotModified;
    this.threadData = threadData;
    this.statusCode = statusCode;
  }

  static success(threadData) {
    return new ThreadFetchResult({ isSuccess: true, threadData });
  }

  static notModified() {
    return new ThreadFetchResult({ isNotModified: true });
  }

  static failed(statusCode) {
    return new ThreadFetchResult({ statusCode });
  }
}

export class ThreadFetchService {
  constructor(fetchImpl = globalThis.fetch) {
    this.fetchImpl = fetchImpl;
  }

  async fetchCatalog(board) {
    const response = await this.fetchImpl(URLS.catalogUrl(board), {
      headers: {
        'User-Agent': USER_AGENT
      }
    });

    if (response.ok) {
      const catalogPages = await response.json();
      return {
        isSuccess: true,
        statusCode: response.status,
        catalogPages: catalogPages.map(createCatalogThreads)
      };
    }

    return {
      isSuccess: false,
      statusCode: response.status,
      catalogPages: []
    };
  }

  async fetchThread(thread) {
    const response = await this.fetchImpl(URLS.threadUrl(thread.Board, thread.ThreadId), {
      headers: {
        'User-Agent': USER_AGENT,
        'If-Modified-Since': formatHttpDate(thread.LastChecked)
      }
    });

    if (response.ok) {
      return ThreadFetchResult.success(createThread(await response.json()));
    }

    if (response.status === 304) {
      return ThreadFetchResult.notModified();
    }

    return ThreadFetchResult.failed(response.status);
  }
}

export function formatHttpDate(value) {
  const date = new Date(value);
  return Number.isNaN(date.getTime()) ? new Date(0).toUTCString() : date.toUTCString();
}
